### PR TITLE
Support bridging services

### LINKED ISSUE #70: Bridge services
Ignition offers several services. It would be good to bridge them directly to ROS (1 and 2) so users can do things like spawning entities directly from ROS.

### PR DESCRIPTION
**Pull request by ivanpauno**
# 🎉 New feature

Closes #70

## Summary

* Add support for bridging services.
  At the moment you can only bridge the world control service.
* Add some message definitions that can be useful for bridging other ignition services.

I plan to clean up the code a bit and support to bridge more services in a follow up PR.

## Test it

```bash
ros ign /path/to/any/world/with/ign/gui/world/control/enabled
ros2 run ros_ign_bridge parameter_bridge /world/wheel_slip/control@ros_ign_interfaces/srv/ControlWorld
ros2 service call /world/<world_name>/control ros_ign_interfaces/srv/ControlWorld "world_control:
  pause: false
  step: false
  multi_step: 0
  reset:
    all: false
    time_only: false
    model_only: false
  seed: 0
  run_to_sim_time:
    sec: 0
    nanosec: 0"
```

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by ahcorde — DISMISSED** [review]
I can't see any tests here, is there any chance to add then ?

**Comment by ivanpauno** [comment]
> I can't see any tests here, is there any chance to add then ?

Yes, I will add tests.
I opened a PR to get some early feedback.

**Comment by ivanpauno** [comment]
> I can't see any tests here, is there any chance to add then ?

See https://github.com/ignitionrobotics/ros_ign/pull/211/commits/ab63ae6a12bfdcd1fe85b0a539f8023b81691338.

File: ros_ign_bridge/include/ros_ign_bridge/convert/geometry_msgs.hpp [hunk 1]
```diff
@@ -15,6 +15,13 @@
 #ifndef ROS_IGN_BRIDGE__CONVERT__GEOMETRY_MSGS_HPP_
 #define ROS_IGN_BRIDGE__CONVERT__GEOMETRY_MSGS_HPP_
 
+#include <ignition/msgs/quaternion.pb.h>
+#include <ignition/msgs/vector3d.pb.h>
+#include <ignition/msgs/pose.pb.h>
+#include <ignition/msgs/pose_v.pb.h>
+#include <ignition/msgs/twist.pb.h>
+#include <ignition/msgs/wrench.pb.h>
```
Comment (chapulina) [inline ros_ign_bridge/include/ros_ign_bridge/convert/geometry_msgs.hpp]: These order changes could be broken into a separate PR to make this one smaller

File: ros_ign_bridge/include/ros_ign_bridge/convert/builtin_interfaces.hpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#include <builtin_interfaces/msg/time.hpp>
+
+#include "ros_ign_bridge/convert_decl.hpp"
+
+namespace ros_ign_bridge
+{
+
+template<>
+void
+convert_ros_to_ign(
```
Comment (chapulina) [inline ros_ign_bridge/include/ros_ign_bridge/convert/builtin_interfaces.hpp]: This conversion is standalone enough that it could come in a separate PR. Then it's easier to check that it has all the proper testing and documentation (for example, it should be listed on `ros_ign_bridge/README`

File: ros_ign_bridge/src/parameter_bridge.cpp [hunk 3]
```diff
@@ -35,22 +34,31 @@ enum Direction
...
     "The ROS message type is followed by an @, [, or ] symbol where\n" <<
     "    @  == a bidirectional bridge, \n" <<
     "    [  == a bridge from Ignition to ROS,\n" <<
     "    ]  == a bridge from ROS to Ignition.\n" <<
     "Following the direction symbol is the Ignition Transport message " <<
     "type.\n\n" <<
+    "Services: The first @ symbol delimits the service name from the types.\n" <<
+    "Following the first @ symbol is the ROS service type.\n" <<
+    "Optionally, you can include the ignition request and reply type\n" <<
+    "separated by the @ symbol.\n" <<
```
Comment (chapulina) [inline ros_ign_bridge/src/parameter_bridge.cpp:59]: Should we mention what it defaults to if the Ignition types are omitted?

Just checking if the Ignition service optional because we're assuming that there's a 1-to-1 mapping. I can't think of a reason for it not to be. In that case, I think we could reduce complexity and even remove the Ignition services as parameters :thinking:

File: ros_ign_bridge/src/parameter_bridge.cpp [hunk 4]
```diff
@@ -35,22 +34,31 @@ enum Direction
...
+    "It is only supported to expose ignition servces as ros services, i.e.\n"
```
Comment (chapulina) [inline ros_ign_bridge/src/parameter_bridge.cpp]:
```suggestion
    "It is only supported to expose Ignition servces as ROS services, i.e.\n"
```

File: ros_ign_bridge/src/parameter_bridge.cpp [hunk 5]
```diff
@@ -89,23 +82,26 @@ int main(int argc, char * argv[])
     usage();
     return -1;
   }
-
-  rclcpp::init(argc, argv);
+  // skip the process name in argument procesing
+  ++argv;
+  --argc;
+  auto filteredArgs = rclcpp::init_and_remove_ros_arguments(argc, argv);
```
Comment (chapulina) [inline ros_ign_bridge/src/parameter_bridge.cpp:89]: This is a nice cleanup! Also a good candidate for its own PR

File: ros_ign_bridge/src/service_factories/ros_ign_interfaces.cpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  ignition::msgs::WorldControl & ign_req)
+{
+  convert_ros_to_ign(ros_req.world_control, ign_req);
+}
+
+template<>
+void
+convert_ign_to_ros(
+  const ignition::msgs::Boolean & ign_rep,
+  ros_ign_interfaces::srv::ControlWorld::Response & ros_res)
```
Comment (chapulina) [inline ros_ign_bridge/src/service_factories/ros_ign_interfaces.cpp:63]: Just an idea for the future. I assume we'll have many of these responses that just hold a `success` boolean.  It would be interesting to abstract that into something like

```cpp
template<BooleanSuccessResponse>
void
convert_ign_to_ros(
  const ignition::msgs::Boolean & ign_rep,
  BooleanSuccessResponse & ros_res)
{
  ros_res.success = ign_rep.data();
}
```

File: ros_ign_bridge/src/service_factory.hpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+#include "service_factory_interface.hpp"
+
+namespace ros_ign_bridge
+{
+
+template<typename RosResT>
+bool
+send_response_on_error(RosResT & ros_response);
+
+template<typename RosServiceT, typename IgnRequestT, typename IgnReplyT>
```
Comment (chapulina) [inline ros_ign_bridge/src/service_factory.hpp:38]:
```suggestion
template<typename RosServiceT, typename IgnRequestT, typename IgnResponseT>
```

"response" is used throughout Ignition Transport's codebase

File: ros_ign_bridge/test/launch/test_ign_server.launch.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+from launch import LaunchDescription
+
+from launch_ros.actions import Node
+
+import launch_testing
+
+
+def generate_test_description():
+
+    publisher = Node(
```
Comment (chapulina) [inline ros_ign_bridge/test/launch/test_ign_server.launch.py]:
```suggestion
    server = Node(
```

File: ros_ign_bridge/test/serverclient/ign_server.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+#include <chrono>
+#include <csignal>
+#include <iostream>
+#include <string>
+#include <thread>
+
+#include "utils/test_utils.hpp"
+#include "utils/ign_test_msg.hpp"
+
+/// \brief Flag used to break the publisher loop and terminate the program.
```
Comment (chapulina) [inline ros_ign_bridge/test/serverclient/ign_server.cpp]:
```suggestion
/// \brief Flag used to break the waiting loop and terminate the program.
```

File: ros_ign_bridge/test/serverclient/ign_server.cpp [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  // Install a signal handler for SIGINT and SIGTERM.
+  std::signal(SIGINT, signal_handler);
+  std::signal(SIGTERM, signal_handler);
+
+  // Create a transport node and advertise a topic.
+  ignition::transport::Node node;
+
+  // ignition::msgs::Color.
+  node.Advertise(
+    "/ign_ros/test/serviceclient/service",
```
Comment (chapulina) [inline ros_ign_bridge/test/serverclient/ign_server.cpp]: It would be good to make this service specific to the request type so we can extend the test later, something like

```suggestion
    "/ign_ros/test/serviceclient/control_world",
```

File: ros_ign_interfaces/msg/VideoRecord.msg [hunk 11]
```diff
@@ -0,0 +1,16 @@
+# A message that allows for control of video recording functions.
```
Comment (chapulina) [inline ros_ign_interfaces/msg/VideoRecord.msg]: > I plan to clean up the code a bit and support to bridge more services in a follow up PR.

I see some of these messages aren't used yet, right? It would be nice to add them together with the PR where they will be used.

**Review by chapulina — COMMENTED** [review]
This is great! The overall approach looks good to me! I mainly have suggestions on how to split up the PR, so that it's easier to check that all new code is tested and documented.

File: ros_ign_bridge/include/ros_ign_bridge/convert/geometry_msgs.hpp (hunk 1 above)
Comment (chapulina) [reply ros_ign_bridge/include/ros_ign_bridge/convert/geometry_msgs.hpp]: @mjcarroll took care of it in #213

**Comment by ivanpauno** [comment]
> I mainly have suggestions on how to split up the PR, so that it's easier to check that all new code is tested and documented.

Sounds good!
Yeah, I ended up putting a lot of different things together.

**Comment by mjcarroll** [comment]
Alright @ivanpauno this is probably ready for a rebase.  Sorry for causing so much chaos with the other fixes, but now we are getting debs again on galactic and rolling.

**Comment by ivanpauno** [comment]
> Alright @ivanpauno this is probably ready for a rebase. Sorry for causing so much chaos with the other fixes, but now we are getting debs again on galactic and rolling.

No problem, that's easy to fix, I only need to discard [one commit](https://github.com/ignitionrobotics/ros_ign/pull/211/commits/10165a147e8e818fe2b25dd166c475158309e6a3) :wink: .
I will probably be opening some PRs split from this one (similar to https://github.com/ignitionrobotics/ros_ign/pull/214), before rebasing.

File: ros_ign_bridge/src/parameter_bridge.cpp (hunk 3 above)
Comment (ivanpauno) [reply ros_ign_bridge/src/parameter_bridge.cpp:59]: > In that case, I think we could reduce complexity and even remove the Ignition services as parameters thinking

That sounds fine to me, I kept that option as currently we specify both types for topics.

File: ros_ign_bridge/src/service_factories/ros_ign_interfaces.cpp (hunk 6 above)
Comment (ivanpauno) [reply ros_ign_bridge/src/service_factories/ros_ign_interfaces.cpp:63]: Yes, that sounds fine to me.

File: ros_ign_bridge/src/service_factory.hpp (hunk 7 above)
Comment (ivanpauno) [reply ros_ign_bridge/src/service_factory.hpp:38]: Actually, [Reply is used](https://github.com/ignitionrobotics/ign-transport/blob/e02bd8b99bdcaff9bae401d791236317dd27c6d3/include/ignition/transport/Node.hh#L371).
In `ign-gazebo` though, there are more instances of "response" than "reply".

If `Response` is more common for ignition developers, I will update it here.
It also matches the ROS 2 naming.

File: ros_ign_bridge/test/serverclient/ign_server.cpp (hunk 10 above)
Comment (ivanpauno) [reply ros_ign_bridge/test/serverclient/ign_server.cpp]: done in https://github.com/ignitionrobotics/ros_ign/pull/211/commits/1b4c9351912033f0a457042fff8b63f3ed059b22

File: ros_ign_bridge/test/serverclient/ign_server.cpp (hunk 9 above)
Comment (ivanpauno) [reply ros_ign_bridge/test/serverclient/ign_server.cpp]: done in https://github.com/ignitionrobotics/ros_ign/pull/211/commits/1b4c9351912033f0a457042fff8b63f3ed059b22

File: ros_ign_bridge/test/launch/test_ign_server.launch.py (hunk 8 above)
Comment (ivanpauno) [reply ros_ign_bridge/test/launch/test_ign_server.launch.py]: done in https://github.com/ignitionrobotics/ros_ign/pull/211/commits/1b4c9351912033f0a457042fff8b63f3ed059b22

File: ros_ign_bridge/test/serverclient/ros_client.cpp [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+/////////////////////////////////////////////////
+TEST(ROSClientTest, WorldControl)
+{
+  rclcpp::init(0, NULL);
+  auto node = std::make_shared<rclcpp::Node>("test_ros_client_to_ign_service");
+  auto client = node->create_client<ros_ign_interfaces::srv::ControlWorld>(
+    "/ign_ros/test/serviceclient/world_control");
+  std::this_thread::sleep_for(1s);
+  client->wait_for_service();
```
Comment (jacobperron) [inline ros_ign_bridge/test/serverclient/ros_client.cpp]: We should probably have a timeout here to avoid blocking forever in case there's an problem.

File: ros_ign_bridge/test/serverclient/ign_server.cpp [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+#include <ignition/msgs/boolean.pb.h>
+#include <ignition/msgs/world_control.pb.h>
+
+#include <ignition/transport.hh>
+
+#include <atomic>
+#include <chrono>
+#include <csignal>
+#include <iostream>
+#include <string>
```
Comment (jacobperron) [inline ros_ign_bridge/test/serverclient/ign_server.cpp]: unused?

File: ros_ign_bridge/src/parameter_bridge.cpp [hunk 14]
```diff
@@ -35,22 +35,31 @@ enum Direction
...
     "    ]  == a bridge from ROS to Ignition.\n" <<
     "Following the direction symbol is the Ignition Transport message " <<
     "type.\n\n" <<
+    "Services: The first @ symbol delimits the service name from the types.\n" <<
+    "Following the first @ symbol is the ROS service type.\n" <<
+    "Optionally, you can include the ignition request and reply type\n" <<
+    "separated by the @ symbol.\n" <<
+    "It is only supported to expose Ignition servces as ROS services, i.e.\n"
+    "the ros service will forward request to the ignition service and then forward\n"
+    "the reply back to the ros client.\n\n"
```
Comment (jacobperron) [inline ros_ign_bridge/src/parameter_bridge.cpp]:
```suggestion
    "the response back to the ROS client.\n\n"
```

File: ros_ign_bridge/src/parameter_bridge.cpp [hunk 15]
```diff
@@ -35,22 +35,31 @@ enum Direction
...
     "    [  == a bridge from Ignition to ROS,\n" <<
     "    ]  == a bridge from ROS to Ignition.\n" <<
     "Following the direction symbol is the Ignition Transport message " <<
     "type.\n\n" <<
+    "Services: The first @ symbol delimits the service name from the types.\n" <<
+    "Following the first @ symbol is the ROS service type.\n" <<
+    "Optionally, you can include the ignition request and reply type\n" <<
+    "separated by the @ symbol.\n" <<
+    "It is only supported to expose Ignition servces as ROS services, i.e.\n"
+    "the ros service will forward request to the ignition service and then forward\n"
```
Comment (jacobperron) [inline ros_ign_bridge/src/parameter_bridge.cpp]:
```suggestion
    "the ROS service will forward request to the Ignition service and then forward\n"
```

File: ros_ign_bridge/test/serverclient/ros_client.cpp (hunk 12 above)
Comment (ivanpauno) [reply ros_ign_bridge/test/serverclient/ros_client.cpp]: Done in https://github.com/ignitionrobotics/ros_ign/pull/211/commits/b2a37daabf00db7c9bfed437d7d096dc535f597a

File: ros_ign_bridge/test/serverclient/ign_server.cpp (hunk 13 above)
Comment (ivanpauno) [reply ros_ign_bridge/test/serverclient/ign_server.cpp]: Fixed in https://github.com/ignitionrobotics/ros_ign/pull/211/commits/b2a37daabf00db7c9bfed437d7d096dc535f597a

File: ros_ign_bridge/src/parameter_bridge.cpp (hunk 14 above)
Comment (ivanpauno) [reply ros_ign_bridge/src/parameter_bridge.cpp]: See https://github.com/ignitionrobotics/ros_ign/pull/211/commits/b2a37daabf00db7c9bfed437d7d096dc535f597a

File: ros_ign_bridge/src/parameter_bridge.cpp (hunk 15 above)
Comment (ivanpauno) [reply ros_ign_bridge/src/parameter_bridge.cpp]: Fixed in https://github.com/ignitionrobotics/ros_ign/pull/211/commits/b2a37daabf00db7c9bfed437d7d096dc535f597a and https://github.com/ignitionrobotics/ros_ign/pull/211/commits/79611258cb788b3957d8e3406fe92b57a41685ba

**Review by jacobperron — APPROVED** [review]
LGTM, but I think we should get someone from the Ignition team to approve ultimately.

**Comment by ivanpauno** [comment]
CI failures seem to be unrelated ...

**Review by chapulina — APPROVED** [review]
Great work! :+1:

**Comment by ivanpauno** [comment]
@chapulina should I wait for the CI (infrastructure?) issue to be fixed or can this be merged?

**Comment by chapulina** [comment]
It would be good to have CI passing to merge this, since it's such a big change. I believe everything here should be backportable, right? How about targeting `foxy` and we merge it forward? That's usually our preferred pattern anyway, and will allow us to run CI for this while #219 is being worked on.

**Comment by ivanpauno** [comment]
Applying the same changes to the foxy branch is not that easy after https://github.com/ignitionrobotics/ros_ign/pull/192 and some other PRs.
I will check what I can do

**Comment by ivanpauno** [comment]
> Applying the same changes to the foxy branch is not that easy after https://github.com/ignitionrobotics/ros_ign/pull/192 and some other PRs.
I will check what I can do

Well, either way it's going to be hard to rebase :cold_sweat:.
@chapulina do you prefer to also have #192 and the related PRs applied to foxy/galactic/etc or should I try to apply this change directly to `foxy`?

In the later case, I probably want to apply https://github.com/ignitionrobotics/ros_ign/pull/214 and https://github.com/ignitionrobotics/ros_ign/pull/216 before.

**Comment by chapulina** [comment]
> do you prefer to also have https://github.com/ignitionrobotics/ros_ign/pull/192 and the related PRs applied to foxy/galactic/etc 

Those aren't easily backportable either because they change public headers :confused: 

> or should I try to apply this change directly to foxy?

If you're not in a rush to get this in, it may be better to wait until we have CI working on Jammy. It's probably not worth the effort trying to backport this right now.

**Comment by ivanpauno** [comment]
> If you're not in a rush to get this in, it may be better to wait until we have CI working on Jammy. It's probably not worth the effort trying to backport this right now.

Sounds good, I think we're not in a rush, and worst case we could use a fork temporally (or complete the rebasing).

**Comment by chapulina** [comment]
No new test failures, merging!
